### ISSUE TITLE
FEATURE: Mark played titles

### ISSUE THREAD

**Issue by mhkull** [mpogue2/SquareDesk#81]
When you do a Festival and you use a playlist with many entries it would be a great help if the played tracks get a marker of some kind - for example a color code.

**Comment by mhkull** [comment]
Addition: pls try to keep up the clean look on mac. Although I ask for more features I always appreciate a limited number of items on screen. It keeps the information in the center of attention. Other programs provide too many buttons on screen.

**Comment by mpogue2** [comment]
Thanks!  We're trying to keep the information overload under control... :-)
I like the idea of marking played tracks, and having the ability to clear the markers all at once.
Should this be a graphic icon of some sort, or a color change for the whole row?  What do other programs do?

**Comment by mhkull** [comment]
The highlight the songs with a row background color (or make em dissapear from the list - a solution I personally do not like at all).

**Comment by mpogue2** [comment]
I like the highlight idea best, too.

Another thought:  we already track Age by Session.  We could make those used with Age == 0 a grey color (dimmed out a bit), and the dimming proportional to how long ago they were used.  Or, the colors could be Red for recent, Yellow for >1 week, Green for >2 weeks, etc.  In any case, tying the shading to the Age would give us Shading by Session, which is a cool idea!

Also:  right now the foreground color is by Type.  The background color(s) for Age would need to be chosen carefully, so that it doesn't make the text unreadable.

**Comment by phoef** [comment]
I will prefer shades of grey. Or make it customizable by the user. Or offering different sets of colors, where the user can choose from...

**Comment by mhkull** [comment]
The solution should be an eyecatcher. I personally like the way SquareDesk is working to prepare a dance already very much. Age is of help. However I am talking about the live evebnt - the Festival with dancers moving, hands shaking and suddenly you do not remember whether you have done the song already or not. I doubt a grey shade will do the job. I prefer more like a red light.

**Comment by mpogue2** [comment]
The red light or grey background would come on when you PLAY the song, right?  And, if you played a song by accident, maybe there would also need to be a way to clear the color (maybe a Right-click context menu, since this feature would not be used often), clearing the "played recently" color until played again?

**Comment by mhkull** [comment]
You are right. Maybe the red light will only come on if the song is played 60 or 90 seconds.
Otherwise the undo option is great. Right click is also fine, for this is something you do during brakes.

**Comment by mpogue2** [comment]
SqView uses a bright green background for played songs.  I'm not quite sure when a SqView song goes back to a normal background color.  The foreground color is always black.

My proposal is that we do a couple of things:
1) Preference:  [X] Change background color for recently used songs, defaults OFF, user color: [ ]
2) Right click context menu option: Clear Age -- changes it to " " (not 0 which means "played today")
3) Modify the songTable view, so that Age of 0-7 (inclusive) days is user-selectable background color, and all others are normal color.  That means that this feature indicates "I played it today or last week".
4) Provide a new Session:  "Festival", to hold the Ages for special events.  (The only non-day-oriented session name is "Practice", which doesn't really seem obvious to me to use for festivals.)
I think no festival goes longer than 1 week, so item 3 above works for festivals, too.
5) Provide a new menu option: Session > Clear Ages for <current session name>..., that when triggered will clear the Ages for all of the songs in the currently selected Session.  It pops up an "Are you sure?" dialog, and if yes, then it clears the Ages (again, to " ", not zero).

What do you think?

**Comment by danlyke** [comment]
So now that song age has been in for a while, what's the general feeling on how that meshes with this feature?

**Comment by mpogue2** [comment]
I think they work together...you can display the Age numerically, or use the color coding.  I'm still for implementing this, as per above (or something similar)...

**Comment by mpogue2** [comment]
To summarize the discussion over at #161:  We need to provide a list of music PLAYED at the event.  SqVIEW has a function for that called "TRACKING".  Just a playlist of songs that might have been planned to be played is not sufficient (so, Playlist > Print... isn't compliant with GEMA).  

We'll take a look at what SqView does....

**Comment by mpogue2** [comment]
See the detailed description of what the GEMA-compliant printed sheet must have as part of #161 (which is just simple printing of the playlist right now).

**Comment by danlyke** [comment]
So where are we with this? Creating a simple tab that displays the play history that someone could copy and paste from seems super easy, and like a big win for mkhull.

**Comment by mpogue2** [comment]
In 0.8.3alpha6, we now have a column called "recent" that can be cleared.  Session > Clear Recent will clear out all the red triangles, and it essentially sets the "fence",  and anything played after that gets marked "recent".  The marker is a red triangle, which indicates "you've played this already".  The column can be hidden or visible, as per your preference.

It works off the Ages column, so this is NOT a replacement for the Ages column.

I think what's missing for GEMA compliance is the printing part, as described in #161 (and repeated below for reference).  Then, going into a dance or a festival, the caller can do a "Session > Clear Recent", call whatever, and then probably something like File > Print Recent, which brings up a dialog with the non-song details, and then spits out a CSV file or a WORD file (or a PDF?).   Perhaps we could also have File > Email Recent, which would make the CSV/WORD file, and then send it to somebody.

**Sample:**
Club: Offenburger SDC
Event: Christmas Party
Date: 2.12.2017
Place: Rammersweier
Time: 14:00 - 22:00 Uhr
Leader: Martin Kull, Max Mustermann, ...

Song Title, Label, Nr, Artist, etc.
Driving Home For Christmas, MM 251, Chris Rea

**Comment by mpogue2** [comment]
Or maybe File > Recent > Print and File > Recent > Email ?

For reference, the Session > Clear Recent and Recent column work was done as part of #193.

**Comment by mhkull** [comment]
Hi all,

I am no developer, so all "code messages" have been ignored by me due to a lack of understanding.

IMHO a simple CSV export is doing the job and the user can open the file in another program and add whatever she/he wants.

**Comment by phoef** [comment]
I like the idea of a complete play history, independently of the age- or recent-feature. A complete history with the following informations: File, Type, Date, Time. Perhaps the selected Session? 

The list should be sortable by date. And a possibility to mark and/or copy a selection of played songs. 

Is it an idea to permanently write the play history to a, for example, csv file in the squaredancemusic directory?

**Comment by danlyke** [comment]
We have this history in the database, we just need to figure out how to
expose it nicely to you!

I could probably whip together a "File->Export History" tool right
fast.

Dan

**Comment by phoef** [comment]
This seems to me to be a pragmatic solution. Anyone can prepare the data in the form they need. I like that ;-)

**Comment by danlyke** [comment]
Okay, I'm *finally* getting around to this. Quick question: What's your usual period for generating the reports? I'm trying to figure out good defaults for from and to dates or otherwise restricting the range.

**Comment by phoef** [comment]
From my point of view from yesterday to today is a useful default value. 
In my case I normally export the list directly after the event, or (mostly) the day after.
If there is a multi-day-event it is no problem to adjust the "from" and "to" value, just as I export the list later.

**Comment by danlyke** [comment]
![squaredesksongplayexport](https://user-images.githubusercontent.com/88919/40433148-2d4719b0-5e60-11e8-95fb-adb9aa144ce8.png)

Default is the last 24 hours (I'm open to whatever, I was just writing it somewhere that I didn't have net connectivity and that's what I could do easily without Qt docs..). Output looks like:

```"Song","when played"
"A Song In The Night - TB-182","2018-04-04 13:14:45"
"MTA - ESP 1119","2018-04-20 14:41:57"
"C 222 - Down at the Twist and Shout (Instrumental)","2018-04-24 18:10:03"
"RWH 1187 - Dream A Little Dream of Me","2018-04-24 18:12:41"
"ACME 125 - Slow Hand","2018-04-24 18:22:59"
"CREST 139 - Beautiful Loser","2018-04-24 18:28:03"
```

First column is the full name of the song file, without the `.mp3` or the path, second column is the local time (we store in UTC, this is deliberately converted) of the play,

Commit 47c756b2ca98e4069447778eddfcbd7408bd4d4d

**Comment by danlyke** [comment]
Additional note: This is under "File->Export Play Data..." and the CSV has a header line of:
`"Song","when played"`
